gh-95914: Add paragraph about PEP 654 in main body of 'What's New in 3.11' (#95937)
python/cpython: Doc/whatsnew/3.11.rst

See :pep:`657` for more details. (Contributed by Pablo Galindo, Batuhan Taskaya

> and Ammar Askar in :issue:`43950`.)
> 
> Exception Groups and ``except*`` (PEP 654)

Reviewer: ```suggestion
and Ammar Askar in :issue:`43950`.)

Exception Groups and ``except*`` (PEP 654)
```

It isn't consistent, but generally two blank lines are currently used before h3 headings in the current document, and in particular that may help reduce merge conflicts between PRs that modify adjacent sections (e.g. I have #95915 open right now), so it's probably best to keep to that

> Exception Groups and ``except*`` (PEP 654)

Reviewer: Could we be consistent about PEP numbers in Feature section titles? Right now the type hint PEP sections are in the form `PEP NNN: <Title>`, this section and the one below is in the form `<Title> (PEP NNN)`, and the two above it don't mention the PEP number at all in the title.

IMO, it would seem to make sense to either elide the PEP numbers in the New Features section titles (as I do for the section below in #95914 , since they don't add much value for most users and are already mentioned/linked in the section body and Summary), or consistently use them with the same format between these two sections (I can take care of the others in my separate PRs, but it would be nice to be consistent here).

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
------------------------------------------
```

Like the heading below it (fixed in #95915), this heading needs to be moved up a level, as currently its a subsection of "Enhanced Error Locations in Tracebacks"

:pep:`654` introduces language features that enable a program
to raise and handle multiple unrelated exceptions simultaneously.
The builtin types :exc:`ExceptionGroup` and :exc:`BaseExceptionGroup`
make it possible to group exceptions and raise them together.

> The new ``except*`` syntax generalizes ``except`` to match subgroups

Reviewer 2: In line with our policy that PEPs are historic documents, not documentation, maybe this should deep link to the part under https://docs.python.org/3.11/reference/compound_stmts.html#try where `except*` is described?

Reviewer 3: Both ``:keyword:`except* <except_star>` `` and ``:ref:`except* <except_star>` `` currently link to the beginning of the `try` section, where the `except_star` label is defined.  For a deep link, the label should be moved down to the relevant paragraph.

Reviewer: Actually, the keyword *appears* to be linked precisely, both judging from [its position in the source](https://github.com/python/cpython/blob/187949ebf2ae36fcf0817a06f4a7637d0a8b7fc5/Doc/reference/compound_stmts.rst?plain=1#L315) on both 3.12 and 3.11, and the fact that the [index entry](https://docs.python.org/3.11/genindex-E.html) correctly [links](https://docs.python.org/3.11/reference/compound_stmts.html#index-12) to it. But maybe a `:keyword:` ref doesn't? 

However, the `:ref:` is indeed not; it would be good to move the non-`try` ones (that presumably don't rely on the section name for default link text, since it doesn't even mention except/else/finally) down to the appropriate passages.

Author: I think a link to the top of the section is fine.

> make it possible to group exceptions and raise them together.
> The new ``except*`` syntax generalizes ``except`` to match subgroups

Reviewer: ```suggestion
make it possible to group exceptions and raise them together,
and the new :keyword:`except* <except_star>` syntax
generalizes :keyword:`except` to match subgroups
```

Use the appropriate semantic roles here, and link them.

Also, it currently reads rather choppy; a comma instead of a period (full stop) improves flow somewhat

of exception groups.

See :pep:`654` for more details.

> (Contributed by Irit Katriel in :issue:`45292`. PEP written by

Reviewer 3: ```suggestion
(Contributed by Irit Katriel in :gh:`89455`. PEP written by
```

Author: This should be done (if at all) in a separate PR where the summary reference to the same issue is updated as well.

I say "if at all" because this is not historically accurate - I did the work under a bpo issue. Are you going to change all references to bpo issues in the docs to the gh issues they were mapped to?

> Irit Katriel, Yury Selivanov and Guido van Rossum.)
> 
> Exceptions can be enriched with notes (PEP 678)

Reviewer: ```suggestion
Irit Katriel, Yury Selivanov and Guido van Rossum.)

Exceptions can be enriched with notes (PEP 678)
```

As above

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :meth:`add_note` method was added to :exc:`BaseException`. It can be
used to enrich exceptions with context information which is not available
at the time when the exception is raised. The notes added appear in the
default traceback.

See :pep:`678` for more details.

> (Contributed by Irit Katriel in :issue:`45607`. PEP written by

Reviewer 3: ```suggestion
(Contributed by Irit Katriel in :gh:`89770`. PEP written by
```

> Zac Hatfield-Dodds.)

Reviewer: Could we make these changes in #95915 instead, since that PR is closer to the scope of the change (copyediting existing entries vs. adding a specific new one), and it will cause a significant merge conflict for whichever PR gets merged second otherwise?

Author: ok, I'll remove it and you can put it there.
